### PR TITLE
matthew recruitment training

### PR DESCRIPTION
**Pull request by promatty**
-Utilises useReducer to manage the state of speed, battery, and weather inputs from the child components
-Reducer function uses previous state (spread operator) and updates the new input
-Errors for validation of inputs are stored in a useState, and determined with useCallback
-Range is calculated in a useCallback as well
-Final result message is determined in a useMemo
-A stateChanger prop is passed to each child component, which updates the useReducer state by calling its dispatch function

### CODE REVIEW

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,137 @@
...
-  const calculateRange = () => {
-    return;
+  const [state, dispatch] = useReducer(reducer, {
+    //useReducer to manage more complex states
+    speed: 0, //the initial state of speed, battery, weather values
+    battery: 0,
+    weather: 50,
+  });
+  const [range, setRange] = useState("");
+  const [errors, setErrors] = useState({ speedError: "", batteryError: "" });
```
Comment (burtonjong) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx:27]: Do you really need these to be state? Think of a different way

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx [hunk 2]
```diff
@@ -1,26 +1,128 @@
...
+  type: "changed_speed" | "changed_battery" | "changed_weather";
+  payload: number;
+}
+
 const App = () => {
-  const calculateRange = () => {
-    return;
-  }
+  const [showResult, setShowResult] = useState(false);
+  const [errors, setErrors] = useState({ speedError: "", batteryError: "" });
```
Comment (brian-ngyn) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx:21]: this doesn't need to be state

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx [hunk 3]
```diff
@@ -1,26 +1,128 @@
...
+
+  //the reducer function, updates the state when input values (speed,battery,weather) are changed
+  const reducer = (state: inputState, action: inputAction) => {
```
Comment (brian-ngyn) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx:24]: usecallback

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx [hunk 4]
```diff
@@ -1,26 +1,128 @@
...
+
+  const [state, dispatch] = useReducer(reducer, {
+    //useReducer to manage more complex states
+    speed: 0, //the initial state of speed, battery, weather values
+    battery: 0,
+    weather: 50,
+  });
+
+  //validate speed and battery inputs
+  const validateInputs = useCallback(() => {
```
Comment (brian-ngyn) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx:46]: not sure if this logic is needed to be within a function, you can calculate the errors and memoize it

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,128 @@
...
         <Header />
-        <form name="simulator" className="flex w-full flex-col items-center">
+        <form
+          onSubmit={calculateRange}
+          className="flex w-full flex-col items-center"
+        >
           <div className="mb-4 flex w-full flex-col items-center gap-y-4">
-            <SpeedInput />
-            <BatteryInput />
+            <SpeedInput //pass a stateChanger prop to each component, causing changes to inputs to call the dispatch function of useReducer
```
Comment (brian-ngyn) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx:93]: inputs aren't controlled, if i call dispatch({ type: "changed_speed", payload: 10000 }) then the input box won't read 10000

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,128 @@
...
+              stateChanger={(value) =>
+                dispatch({ type: "changed_speed", payload: value })
+              }
+            />
+            {errors.speedError && (
+              <p className="text-red-500">{errors.speedError}</p>
+            )}
+            <BatteryInput
+              stateChanger={(value) =>
```
Comment (brian-ngyn) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx:102]: same comment as above

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx [hunk 7]
```diff
@@ -1,26 +1,128 @@
...
+            />
+            {errors.batteryError && (
+              <p className="text-red-500">{errors.batteryError}</p>
+            )}
           </div>
           <div className="flex w-full flex-row justify-center gap-4">
-            <WeatherInput />
+            <WeatherInput
+              stateChanger={(value) =>
+                dispatch({ type: "changed_weather", payload: value })
```
Comment (brian-ngyn) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx:113]: same comment as above

File: Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/components/batteryInput/index.tsx [hunk 8]
```diff
@@ -1,4 +1,8 @@
-const BatteryInput = () => {
+interface inputProps {
```
Comment (brian-ngyn) [inline Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/components/batteryInput/index.tsx:1]: Can you reuse this interface and use Pascal case?

**Pull request closed without merging**